File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // halfopen [is, ie)
    TreeNode* helper(vector<int>& preorder, vector<int>& inorder, int pre, int is, int ie){
        for(int i = is; i < ie; i++){
            if (inorder[i] == preorder[pre]){
                TreeNode* root = new TreeNode(preorder[pre]);
                // inorder between [is, i) and prepos > pre
                int prepos = pre;
                root->left = helper(preorder, inorder, prepos + 1, is, i);
                prepos = pre + i - is;

                // inorder between (i, ie) and prepos > pre
                root->right = helper(preorder, inorder, prepos + 1, i + 1, ie);

                return root;
            }
        }
        return nullptr;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return helper(preorder, inorder, 0, 0, inorder.size());
    }
};

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        self.glob = k
        def helper(root):
            if root:
                if self.glob > 0:
                    l = helper(root.left)
                    if l >= 0:
                        return l

                self.glob -= 1
                if not self.glob:
                    return root.val
                
                if self.glob > 0:
                    r = helper(root.right)
                    if r >= 0:
                        return r
            return -1
        return helper(root)